### DISCUSSION TITLE
Points-to-grid: simple multi indexing algorithm for Bloc Sparse

### DISCUSSION THREAD

**Discussion by ArmanddeCacqueray** [getkeops/keops#447] [General]
Hi Keops,

I'm currently doing an internship working on problems where the central issue is computing local Gaussian interactions between a large number of points (you can get a sense of the context [here](https://github.com/For-a-few-DPPs-more/rgbn), though that's not the main point).

As you've noted, Gaussian kernels are insanely sparse — but taking advantage of that sparsity to go beyond the bruteforce approach efficiently on both GPU and CPU turned out to be a real headache for me. After digging into it, I finally stumbled on a workaround that solved my bottlenecks, and I figured it was worth sharing in case it's interesting to you! It felt general enough that it could either serve as a neat alternative for block-sparse operations or maybe spark some new ideas for your side.

The core idea

Since hardware loves tensors, I tried reformulating the irregular neighbor lookup into a tensor problem. The trick was to map the irregular point cloud onto a regular grid,  with a simple greedy algorithm named [Cartesian-Grid-Sort](https://github.com/Space-filling-net/Cartesian-Grid-Sort), that I already implemented in a high level python package (SquareNet) . SquareNet litterally transform the irregular point cloud into a nice, contiguous tensor: each point $(x,y,z)$ gets a coherent spatial multi-index $[i,j,k]$, and those indices together define a regular grid. The idal solution for that would be Optimal Transport, but OT would be too slow for the number of points I consider (up to millions of points). squarenet is based on a very simple heuristic based on 1D sorts which do the job quite well.

Once dataset is structured as a tensor, neighbor lookup becomes a simple matter of shifting indexes — a classic stencil operation that I can delegate to mature stencil libraries (Here Taichi).

It doesn't fit natively into LazyTensors right now because of course their is no way to give a Lazy Tensor stencil instructions natively (without falling back to the standard cluster + block sparse strategy) so I relyed on taichi for the following exemple, but I’d be thrilled if this could inspire ways to integrate similar ideas into KeOps!

Here some visual exemples of how the tensor mapping transform different point clouds, some with holes, peaks... the sweatspot of SquareNet is 2D/3D but it also work in arbitrary dimension.
<img width="705" height="754" alt="squarenet_exemples" src="https://github.com/user-attachments/assets/09aae492-b79a-476b-ad10-edc188d42f53" />

## Benchmark results : Gaussian gradient computations on a 160k point cloud.

N = 400² = 160 000 points, uniform disk, σ² = 6.25×10⁻⁶, device = CUDA

| Method | Build | JIT | Compute | Rel. error |
|---|---|---|---|---|
| V1 Dense KeOps | — | 449 ms | 372 ms | — |
| V2 Block-sparse KeOps | 37 ms | 166 ms | 166 ms | 0.39 * |
| V3 SquareNet + Taichi | 1 268 ms | 78 ms | **3.3 ms** | 1.1×10⁻⁵ |

Result:  **113× faster than dense KeOps**, 50× faster than block-sparse KeOps.

Note on performance trade-offs: The build time for the full SquareNet + Taichi kernel is around 1.3s, which makes it prohibitive if the point cloud is only used once. However, the overhead pays off very quickly after just a few reuses.

## Code

<details>
<summary>Utilities (timing, cluster setup, stencil kernel)</summary>

```python
"""Utility helpers — timing, cluster setup, stencil kernel."""

import math, time
import numpy as np
import torch
import taichi as ti
from pykeops.torch.cluster import (
    grid_cluster, cluster_ranges_centroids, sort_clusters, from_matrix,
)
from squarenet import SquareNet

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
ti.init(arch=ti.gpu)
torch.manual_seed(42)


# ── Dataset ───────────────────────────────────────────────────────────────────

def random_point_cloud(N):
    """Uniform distribution over unit disk."""
    x0 = torch.randn(N, 2, device=DEVICE)
    x0 = x0 / torch.linalg.norm(x0, dim=1, keepdim=True)
    x0 = x0 * torch.rand(N, 1, device=DEVICE) ** (1 / 2)
    return x0


# ── Device sync ───────────────────────────────────────────────────────────────

def sync():
    """Block until all CUDA kernels have finished (no-op on CPU)."""
    if torch.cuda.is_available():
        torch.cuda.synchronize()


# ── V2 build: cluster ranges ──────────────────────────────────────────────────

def build_v2_ranges(x, N, cutoff, device):
    eps    = math.sqrt(100 / N)
    labels = grid_cluster(x, eps)
    xs, _  = sort_clusters(x, labels)

    perm      = sort_clusters(torch.arange(N, device=device), labels)[0]
    inv       = torch.empty_like(perm)
    inv[perm] = torch.arange(N, device=device)

    ranges, centroids, _ = cluster_ranges_centroids(xs, labels)
    D2   = ((centroids[:, None] - centroids[None]) ** 2).sum(-1)
    keep = D2 < D2.median() * cutoff

    return xs, inv, from_matrix(ranges, ranges, keep)


# ── V3 build: SquareNet ───────────────────────────────────────────────────────

def build_v3_squarenet(x, side):
    sq = SquareNet(gridshape=(side, side), backend="torch", verbose=0, max_iter=200)
    sq.fit(x, method="ultimate")
    return sq


# ── V3 Taichi stencil ─────────────────────────────────────────────────────────

@ti.kernel
def _stencil(pts:   ti.types.ndarray(ndim=2),
             grads: ti.types.ndarray(ndim=2),
             sig2: ti.f32, side: ti.i32, wr: ti.i32):
    """GPU stencil: Gaussian gradient over a disk neighbourhood on a 2-D grid.

    For each grid cell (i, j), accumulates g = Σ_neighbours ∂K/∂x where
    K(x, y) = exp(-||x - y||² / sig2). Neighbours are defined by a disk of
    radius wr in grid cells (||di, dj||₂ ≤ wr), clipped at grid borders.
    """
    for idx in range(side * side):
        i, j   = idx // side, idx % side
        xi, yi = pts[idx, 0], pts[idx, 1]
        gx = gy = 0.0
        for di in range(-wr, wr + 1):
            for dj in range(-wr, wr + 1):
                if di ** 2 + dj ** 2 > wr ** 2: continue
                ni, nj = i + di, j + dj
                if 0 <= ni < side and 0 <= nj < side:
                    nidx   = ni * side + nj
                    dx, dy = xi - pts[nidx, 0], yi - pts[nidx, 1]
                    d2     = dx * dx + dy * dy
                    f      = (-2.0 / sig2) * ti.exp(-d2 / sig2)
                    gx    += f * dx
                    gy    += f * dy
        grads[idx, 0] = 2.0 * gx
        grads[idx, 1] = 2.0 * gy


# ── Benchmark runner ──────────────────────────────────────────────────────────

def run_benchmark(x0, grad_v1, compute_v2, compute_v3,
                  build_v2, build_v3, sync_fn):
    reps = 5
    results = {}

    # V1
    sync_fn(); t0 = time.perf_counter()
    grad_v1(x0.clone().requires_grad_(True)); sync_fn()
    t_jit = (time.perf_counter() - t0) * 1e3

    sync_fn(); t0 = time.perf_counter()
    for _ in range(reps):
        g1 = grad_v1(x0.clone().requires_grad_(True))
    sync_fn()
    results["V1 Dense KeOps"] = (0., t_jit, (time.perf_counter() - t0) / reps * 1e3, 0.)

    # V2
    t0 = time.perf_counter()
    xs, inv, rij = build_v2(x0)
    t_build = (time.perf_counter() - t0) * 1e3

    sync_fn(); t0 = time.perf_counter()
    compute_v2(x0, xs, inv, rij); sync_fn()
    t_jit = (time.perf_counter() - t0) * 1e3

    sync_fn(); t0 = time.perf_counter()
    for _ in range(reps):
        g2 = compute_v2(x0, xs, inv, rij)
    sync_fn()
    err2 = (g2 - g1).norm() / g1.norm()
    results["V2 Block-sparse"] = (t_build, t_jit, (time.perf_counter() - t0) / reps * 1e3, err2.item())

    # V3
    t0 = time.perf_counter()
    sq = build_v3(x0)
    t_build = (time.perf_counter() - t0) * 1e3

    sync_fn(); t0 = time.perf_counter()
    compute_v3(x0, sq); sync_fn()
    t_jit = (time.perf_counter() - t0) * 1e3

    sync_fn(); t0 = time.perf_counter()
    for _ in range(reps):
        g3 = compute_v3(x0, sq)
    sync_fn()
    err3 = (g3 - g1).norm() / g1.norm()
    results["V3 SquareNet+Taichi"] = (t_build, t_jit, (time.perf_counter() - t0) / reps * 1e3, err3.item())

    return results
```

</details>

<details>
<summary>Benchmark</summary>

```python
"""
Benchmarking Gaussian gradient computations on a 160k point cloud.

Three strategies:
  V1  Dense KeOps        — LazyTensor, full N×N kernel
  V2  Block-sparse KeOps — grid-clustered, pair-filtered ranges
  V3  Taichi stencil     — space-filling-net layout via SquareNet
"""

N      = 400 ** 2
x0     = random_point_cloud(N)
SIDE   = int(math.sqrt(N))
WR     = 10
CUTOFF = 0.3
SIGMA2 = 1.0 / N
REPS   = 5


def grad_v1(x):
    xi = LazyTensor(x[:, None, :])
    xj = LazyTensor(x[None, :, :])
    K  = (-((xi - xj) ** 2).sum(-1) / SIGMA2).exp()
    return torch.autograd.grad(K.sum(1).sum(), x)[0]


def compute_v2(x, xs_sorted, inv, rij):
    xs = xs_sorted.detach().requires_grad_(True)
    xi = LazyTensor(xs[:, None, :])
    xj = LazyTensor(xs[None, :, :])
    K  = (-((xi - xj) ** 2).sum(-1) / SIGMA2).exp()
    K.ranges = rij
    g  = torch.autograd.grad(K.sum(1).sum(), xs)[0]
    return g[inv]


def compute_v3(x, sq):
    pts   = sq.map(x).detach().cpu().numpy().reshape(-1, 2)
    grads = torch.zeros_like(x).cpu().numpy()
    _stencil(pts, grads, SIGMA2, SIDE, WR)
    return torch.from_numpy(grads).to(DEVICE)[sq.invgridflat.long()]


results = run_benchmark(
    x0         = x0,
    grad_v1    = grad_v1,
    compute_v2 = compute_v2,
    compute_v3 = compute_v3,
    build_v2   = lambda x: build_v2_ranges(x, N, CUTOFF, DEVICE),
    build_v3   = lambda x: build_v3_squarenet(x, SIDE),
    sync_fn    = sync,
)
```

</details>
